### PR TITLE
Chapter 6 edits

### PR DESCRIPTION
**Pull request by dankamongmen**
Some notes:

* Listing 6-2: please don't cast the return value of `malloc()`. this is a comp.lang.c FAQ:  https://c-faq.com/malloc/mallocnocast.html .
If the code is C++, it shouldn't be using `malloc()`.

* 6.1.2: the reason there are no branch mispredictions in the SHA code is because cryptography code must carefully guard against dynamic branching/cache behavior to defend against timing attacks. fascinating stuff. check out DJB's papers.

* 6.1.4 (pg 108)  is section 4.11 really the one you want to reference here? i'm not sure...?

* recommendations to check things with `dmesg` are pretty bad imho. `dmesg` dumps a ring buffer into which the kernel prints over its lifetime. different output settings can filter messages from hitting there. what you almost certainly want is `cpuid`, `cat /proc/cpuinfo`, or `rdmsr`. furthermore, `dmesg` output is in no way a stable api, and it's not available to regular users depending on sysctls.

### CODE REVIEW

File: chapters/6-CPU-Features-For-Performance-Analysis/6-7 Precise Event Based Sampling (PEBS).md [hunk 1]
```diff
@@ -4,15 +4,15 @@ Major CPU vendors provide a set of additional features to enhance performance an
 
 ### PEBS on Intel Platforms {#sec:secPEBS}
 
-Similar to the Last Branch Record feature, PEBS is used while profiling the program to capture additional data with every collected sample. When a performance counter is configured for PEBS, the processor saves the set of additional data, which has a defined format and is called the PEBS record. The format of a PEBS record for the Intel Skylake CPU is shown in Figure @fig:PEBS_record. The record contains the state of general-purpose registers (`EAX`, `EBX`, `ESP`, etc.), `EventingIP`, `Data Linear Address`, and `Latency value`, which will discuss later. The content layout of a PEBS record varies across different microarchitectures, see [@IntelOptimizationManual, Volume 3B, Chapter 20 Performance Monitoring].
```
Comment (dankamongmen) [inline chapters/6-CPU-Features-For-Performance-Analysis/6-7 Precise Event Based Sampling (PEBS).md:7]: sneaky will-vs-we'll! boom

File: chapters/6-CPU-Features-For-Performance-Analysis/6-4 TMA-ARM.md [hunk 2]
```diff
@@ -1,14 +1,14 @@
 ### TMA On ARM Platforms
 
-ARM CPU architects also have developed a TMA performance analysis methodology for their processors, which we will discuss next. ARM calls it "Topdown" in their documentation [@ARMNeoverseV1TopDown], so we will use their naming. At the time of writing this chapter (late 2023), Topdown is only supported on cores designed by ARM, e.g. Neoverse N1 and Neoverse V1, and their derivatives, e.g. Ampere Altra and AWS Graviton3. Refer to the list of major CPU microarchitectures at the end of this book if you need to refresh your memory on ARM chip families. Processors designed by Apple don't support the ARM Topdown performance analysis methodology yet.
```
Comment (dankamongmen) [inline chapters/6-CPU-Features-For-Performance-Analysis/6-4 TMA-ARM.md:3]: what you have here works, but it violates a very common idiom and sounds weird

File: chapters/6-CPU-Features-For-Performance-Analysis/6-4 TMA-ARM.md [hunk 3]
```diff
@@ -67,7 +67,7 @@ Stage 2 (uarch metrics)
...
 
-We don't have a goal of optimizing the benchmark, rather we want to characterize performance bottlenecks. However, if given such a task, here is how our analysis could continue. There is a substantial number of `L1 Data TLB` misses (3.8 MPKI), but then 90% of those misses hit in L2 TLB (see `L2 Unified TLB Miss Ratio`). All in all, only 0.1% of all TLB misses result in a page table walk (see `DTLB Walk Ratio`), which suggests that it is not our primary concern, although a quick experiment that utilizes huge pages is still worth it.
+We don't have a goal of optimizing the benchmark, rather we want to characterize performance bottlenecks. However, if given such a task, here is how our analysis could continue. There are a substantial number of `L1 Data TLB` misses (3.8 MPKI), but then 90% of those misses hit in L2 TLB (see `L2 Unified TLB Miss Ratio`). All in all, only 0.1% of all TLB misses result in a page table walk (see `DTLB Walk Ratio`), which suggests that it is not our primary concern, although a quick experiment that utilizes huge pages is still worthwhile.
```
Comment (dankamongmen) [inline chapters/6-CPU-Features-For-Performance-Analysis/6-4 TMA-ARM.md:70]: definite number disagreement

**Comment by dendibakh** [comment]
> Some notes:
> 
> * Listing 6-2: please don't cast the return value of `malloc()`. this is a comp.lang.c FAQ:  https://c-faq.com/malloc/mallocnocast.html .
>   If the code is C++, it shouldn't be using `malloc()`.

Thanks. Will use `new` instead of `malloc`.

> * 6.1.2: the reason there are no branch mispredictions in the SHA code is because cryptography code must carefully guard against dynamic branching/cache behavior to defend against timing attacks. fascinating stuff. check out DJB's papers.

Good to know.

> * 6.1.4 (pg 108)  is section 4.11 really the one you want to reference here? i'm not sure...?

Ok, I think I meant section 4.9

> * recommendations to check things with `dmesg` are pretty bad imho. `dmesg` dumps a ring buffer into which the kernel prints over its lifetime. different output settings can filter messages from hitting there. what you almost certainly want is `cpuid`, `cat /proc/cpuinfo`, or `rdmsr`. furthermore, `dmesg` output is in no way a stable api, and it's not available to regular users depending on sysctls.

It's been a long time since I looked at it, but `cpuid` may not be enough. It will show you static information (decoded from `CPUID`), but `dmesg` will also show if it is supported by the kernel. I didn't change it from the first edition. I think I borrowed it from some docs that recommended to check using `dmesg`. But yes, for long-running systems, required messages may not be in the buffer. That's a problem.

File: chapters/6-CPU-Features-For-Performance-Analysis/6-4 TMA-ARM.md (hunk 2 above)
Comment (dendibakh) [reply chapters/6-CPU-Features-For-Performance-Analysis/6-4 TMA-ARM.md:3]: Please elaborate on what is wrong and suggest how I can change it.

File: chapters/6-CPU-Features-For-Performance-Analysis/6-4 TMA-ARM.md (hunk 3 above)
Comment (dendibakh) [reply chapters/6-CPU-Features-For-Performance-Analysis/6-4 TMA-ARM.md:70]:
```suggestion
We don't have a goal of optimizing the benchmark, rather we want to characterize performance bottlenecks. However, if given such a task, here is how our analysis could continue. There are a substantial number of `L1 Data TLB` misses (3.8 MPKI), but then 90% of those misses hit in L2 TLB (see `L2 Unified TLB Miss Ratio`). All in all, only 0.1% of all TLB accesses result in a page table walk (see `DTLB Walk Ratio`), which suggests that it is not our primary concern, although a quick experiment that utilizes huge pages is still worthwhile.
```

**Review by dendibakh — APPROVED** [review]

**Pull request merged**